docs/content/docs/setup/configuration.md
---
title: 02 - Configuration
---
Configuration of the app is done through environment variables. See the below options:

> TIP A secret can be generated using: `openssl rand -base64 32`

| Key | Description | Default |
|:----|:------------|:--------|
| `BIND__HOST` | What interface to listen on | `127.0.0.1` |
| `BIND__PORT` | What port to listen on | `8080` |
| `OIDC__ENABLED` | Whether OpenID/OAuth2 Is Enabled | `false` |
| `OIDC__NAME` | The provider name (used for UI) | - |
| `OIDC__ISSUER_URL` | The OIDC issuer url | - |
| `OIDC__CLIENT_ID` | The client id | - |
| `OIDC__CLIENT_SECRET` | The client secret | - |
| `PUBLIC_URL` | Public URL where service can be accessed | - |
| `BEHIND_PROXY` | Whether app is behind a reverse proxy | `false` |
| `DB_URI` | URI for database connection | - |
| `ATTACHMENTS_PATH` | where paste attachments will be stored | - |
| `AUTH_TOKEN_SECRET` | base64 encoded secret  (must be at least 32 bytes) | - |
| `AUTH_TOKEN_EXPIRY` | seconds until a token expires | `604800` |
| `SESSION_SECRET` | base64 encoded secret  (must be at least 32 bytes) | - |
| `SIGNUP_ENABLED` | Whether to allow new accounts to be created | `true` |
| `INTERNAL_AUTH_ENABLED` | Whether to allow login for internal accounts | `true` |
| `RANDOM_SLUG_LENGTH` | How long the randomly generated slug will be | `10` |
| `ANONYMOUS_PASTES_ENABLED` | Whether to allow anonymous users to create new pastes | `true` |
| `MAX_PASTE_SIZE` | Max paste size in bytes including all attachments | `12582912` |
| `ATTACHMENTS_ENABLED` | Whether to allow attachments | `true` |
| `EXPIRY__ANONYMOUS__ENABLED` | Whether to enable default expiry for anonymous users | `false` |
| `EXPIRY__ANONYMOUS__LIMIT_ENABLED` | Whether to force an expiry for anonymous users | `false` |
| `EXPIRY__ANONYMOUS__YEARS` | | `0` |
| `EXPIRY__ANONYMOUS__MONTHS` | | `0` |
| `EXPIRY__ANONYMOUS__DAYS` | | `2` |
| `EXPIRY__ANONYMOUS__HOURS` | | `0` |
| `EXPIRY__USERS__ENABLED` | Whether to enable default expiry for users with accounts | `false` |
| `EXPIRY__USERS__LIMIT_ENABLED` | Whether to force an expiry for users with accounts | `false` |
| `EXPIRY__USERS__YEARS` | | `0` |
| `EXPIRY__USERS__MONTHS` | | `0` |
| `EXPIRY__USERS__DAYS` | | `2` |
| `EXPIRY__USERS__HOURS` | | `0` |

## OIDC
Single-Sign-On is handled via OpenID Connect and OAuth2. To use SSO you must have a compatible provider that supports the following features:

- OpenID Connect (OIDC) Discovery - RFC5785
- Claims
    - `sub`: the users id
    - `name`: a users full name
    - `preferred_username`: the users username, not the email
- Scopes
    - openid
    - profile

Depending on your SSO provider the issuer URL may be different, see below for examples:

Authentik:

```text
https://{example.com}/application/o/{hasty-paste}/
```

## Database URI
Database URIs have to be set in a specific format, see below examples:

SQLite:

```text
sqlite://path/to/db.sqlite
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.4.0] - 2026-05-04
### Added
- Closes #151; enable line numbers for source code pastes
- Paste creation via quick-api
- Cache headers to attachments
### Changed
- Require auth secrets to be at least 32 bytes long
- Remove emojis from index page
- Bump deps

## [2.3.1] - 2026-03-29
### Changed
- Switch to node v24
- Bump deps

## [2.3.0] - 2025-10-10
### Added
- #138; User can delete their own pastes and account
### Changed
- Cleanup now removes deleted accounts and pastes
- Require Node V22 and Go V1.25

## [2.2.0] - 2025-08-18
### Added
- #127; more expiry controls
- admin cli (part of #120)
### Changed
- match ui with Note Mark
### Fixed
- fix navbar going off screen on mobile

## [2.1.0] - 2025-04-24
### Added
- #121; ability to disable anonymous paste creation
- #122; paste size limit & ability to disable attachments
- #123; ability to change random slug length
- #124; docker health check
- #129; migrate lexer alias code to official method
### Changed
- show paste link on paste page
- bump deps
- update to go1.24

## [2.0.0] - 2025-02-27
### Added
- V2 app
### Removed
- V1 app

## [1.10.0] - 2024-02-17
### Changed
- Bump deps & migrate some code
- Use hatch for python management

## [1.9.0] - 2023-02-16
### Added
- #27; :construction: Experimental S3 storage support
- Add ability to hide ascii art boot message
### Changed
- #70; Long ID only adjustable through config value
- Bump deps

## [1.8.0] - 2022-12-21
### Added
- Ensure Redis can be reached on app launch
- Add ASCII art for launch :)
- #67; log loaded configs on launch
- #66; custom 404 page
- Support installing using pip through setuptools pyproject
- Added optional json accelerator (installed by default in Docker image)
### Changed
- #68; make Docker image rootless (increase security)
- Reduce number of final Docker image layers
- Remove ability to use "Copy Share Link" when not under secure connection
- Improved launch scripts
- Bump pip versions
### Fixed
- Fix long id

## [1.7.0] - 2022-10-30
### Added
- Show expiry on paste screen
- Configurable multi-tiered caching
- CLI can remove empty cache folders
- Strict paste id checking in URLs
- Better general exception handling
- Add human padding for paste id
- Some more unit tests
### Changed
- Major code refactoring (pastes are no longer dependant on storage types, for future s3 object support)
- Tidy REST API routes
### Fixed
- Fixed #53 Expiry set in UI always interpreted as UTC, by having a configurable timezone
### Removed
- Removed deprecated features, see #50

## [1.6.1] - 2022-10-09
### Fixed
- #51, When override lexer is given in paste url incorrectly uses cached content

## [1.6.0] - 2022-10-08
### Added
- Simple paste creation through the api
- Ability to hide version number
- Optional caching with internal or redis
### Changed
- Code refactoring
- Update requirements
- Dependency updates

## [1.5.0] - 2022-09-23
### Added
- #39, brand customisation
### Changed
- Only generate paste id's with A-Za-z0-9
- Put config into groups
- Update pip requirements
  - quart-schema to ~=0.13.0
  - pydantic[dotenv] to ~=1.10.2
  - mkdocs-material to ~=8.5.3
- Split routes into separate modules

## [1.4.0] - 2022-08-27
### Added
- #21, Experimental management CLI
### Changed
- #19, Use --link for dockerfile, improving build speed
### Fixed
- #26, Backwards compatibility for copy share link button

## [1.3.0] - 2022-08-19
### Added
- #18, add ability to clone a paste content into new paste
- #20, add optional paste title
- #22, override lexer name with `.<ext>` of paste URL
### Fixed
- Fixed not being able to select "No Highlighting" option

## [1.2.1] - 2022-08-14
### Fixed
- Fixed not being able to select "No Highlighting" option

## [1.2.0] - 2022-08-13
### Added
- #8, add filtered highlighter syntax dropdown
- #6, config option to turn index page into new paste page
### Changed
- add ability to hide long id form checkbox
- #10, centred horizontal positioned options in new paste form
- use shorter names for syntax highlighting
### Fixed
- #9, turn off auto correction features

## [1.1.0] - 2022-08-04
### Added
- Configurable default expiry for UI (disabled by default)
- Optional syntax highlighting
- Optional public list api route (disabled by default)

## [1.0.0] - 2022-07-31
### Added
- Initial release


app/core/config.go
package core

import (
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"github.com/caarlos0/env/v11"
	"github.com/labstack/gommon/bytes"
)

type Base64Decoded []byte

func (b *Base64Decoded) UnmarshalText(text []byte) error {
	decoded, err := base64.StdEncoding.DecodeString(string(text))
	if err != nil {
		return errors.New("cannot decode base64 string")
	}
	*b = decoded
	return nil
}

type Bytes int64

func (b *Bytes) UnmarshalText(text []byte) error {
	if v, err := bytes.Parse(string(text)); err != nil {
		return err
	} else {
		*b = Bytes(v)
		return nil
	}
}

type BindConfig struct {
	Host string `env:"HOST" envDefault:"127.0.0.1"`
	Port uint   `env:"PORT" envDefault:"8080"`
}

func (c *BindConfig) AsAddress() string {
	return fmt.Sprintf("%s:%d", c.Host, c.Port)
}

type DevConfig struct {
	Enabled      bool   `env:"ENABLED" envDefault:"false"`
	ViteDevHost  string `env:"VITE_DEV_HOST,notEmpty" envDefault:"localhost:5173"`
	ViteDistPath string `env:"VITE_DIST_PATH,notEmpty" envDefault:"./dist"`
}

type OIDCConfig struct {
	Enabled      bool   `env:"ENABLED" envDefault:"false"`
	Name         string `env:"NAME"`
	IssuerUrl    string `env:"ISSUER_URL"`
	ClientID     string `env:"CLIENT_ID"`
	ClientSecret string `env:"CLIENT_SECRET"`
}

type ExpiryTimeConfig struct {
	Enabled      bool `env:"ENABLED" envDefault:"false"`
	LimitEnabled bool `env:"LIMIT_ENABLED" envDefault:"false"`
	Years        int  `env:"YEARS" envDefault:"0"`
	Months       int  `env:"MONTHS" envDefault:"0"`
	Days         int  `env:"DAYS" envDefault:"2"`
	Hours        int  `env:"HOURS" envDefault:"0"`
}

type ExpiryConfig struct {
	Anonymous ExpiryTimeConfig `envPrefix:"ANONYMOUS__"`
	Users     ExpiryTimeConfig `envPrefix:"USERS__"`
}

type AppConfig struct {
	Dev                    DevConfig     `envPrefix:"DEV__"`
	Bind                   BindConfig    `envPrefix:"BIND__"`
	OIDC                   OIDCConfig    `envPrefix:"OIDC__"`
	PublicURL              string        `env:"PUBLIC_URL,notEmpty"`
	BehindProxy            bool          `env:"BEHIND_PROXY" envDefault:"false"`
	DbUri                  string        `env:"DB_URI,notEmpty"`
	AttachmentsPath        string        `env:"ATTACHMENTS_PATH,notEmpty"`
	TokenSecret            Base64Decoded `env:"AUTH_TOKEN_SECRET,notEmpty"`
	TokenExpiry            int64         `env:"AUTH_TOKEN_EXPIRY" envDefault:"604800"`
	SessionSecret          Base64Decoded `env:"SESSION_SECRET,notEmpty"`
	SignupEnabled          bool          `env:"SIGNUP_ENABLED" envDefault:"true"`
	InternalAuthEnabled    bool          `env:"INTERNAL_AUTH_ENABLED" envDefault:"true"`
	RandomSlugLength       int           `env:"RANDOM_SLUG_LENGTH" envDefault:"10"`
	AnonymousPastesEnabled bool          `env:"ANONYMOUS_PASTES_ENABLED" envDefault:"true"`
	MaxPasteSize           int64         `env:"MAX_PASTE_SIZE" envDefault:"12582912"`
	AttachmentsEnabled     bool          `env:"ATTACHMENTS_ENABLED" envDefault:"true"`
	Expiry                 ExpiryConfig  `envPrefix:"EXPIRY__"`
}

func (ac *AppConfig) SecureMode() bool {
	return strings.HasPrefix(ac.PublicURL, "https://")
}

func (appConfig *AppConfig) ParseConfig() error {
	if err := env.Parse(appConfig); err != nil {
		return err
	}
	if len(appConfig.TokenSecret) < 32 {
		return fmt.Errorf("AUTH_TOKEN_SECRET length must be at least 32, got %d", len(appConfig.TokenSecret))
	}
	if len(appConfig.SessionSecret) < 32 {
		return fmt.Errorf("SESSION_SECRET length must be at least 32, got %d", len(appConfig.SessionSecret))
	}
	return nil
}
